crates/ra_hir/src/util.rs

```diff
@@ -0,0 +1,19 @@
+//! Internal utility functions.
+
+use std::sync::Arc;
+
+/// Helper for mutating `Arc<[T]>` (i.e. `Arc::make_mut` for Arc slices).
+/// The underlying values are cloned if there are other strong references.
+pub(crate) fn make_mut_arc_slice<T: Clone, R>(
+    a: &mut Arc<[T]>,
+    f: impl FnOnce(&mut [T]) -> R,
+) -> R {
```

Comment:
What do you think about changing this to

```rust
fn make_mut_slice(&mut Arc<[T]>) -> &mut [T]
```

This way, it'll more closely resamble `Arc::make_mut`  in signature, and the function won't be generic, which might slightly improve compile times in the long run

Reply:
I tried that way initially but couldn't pass borrowck.

```rust
pub fn make_mut_arc_slice<T: Clone>(
    a: &mut Arc<[T]>,
) -> &mut [T] {
    if let Some(s) = Arc::get_mut(a) {
        return s;
    }

    *a = a.iter().cloned().collect();
    Arc::get_mut(a).unwrap()
}
```

```
error[E0506]: cannot assign to `*a` because it is borrowed
 --> src/main.rs:9:5
  |
3 |     a: &mut Arc<[T]>,
  |        - let's call the lifetime of this reference `'1`
4 | ) -> &mut [T] {
5 |     if let Some(s) = Arc::get_mut(a) {
  |                                   - borrow of `*a` occurs here
6 |         return s;
  |                - returning this value requires that `*a` is borrowed for `'1`
...
9 |     *a = a.iter().cloned().collect();
  |     ^^ assignment to borrowed `*a` occurs here
```

Reply:
`Arc::make_mut` touches `Arc`'s internals, so I guess the same thing cannot be implemented outside std.

Reply:
Don't know why, but the following pacifies the borrowch:

```rust
use std::sync::Arc;

pub fn make_mut_arc_slice<T: Clone>(
    a: &mut Arc<[T]>,
) -> &mut [T] {
    if Arc::get_mut(a).is_none() {
        *a = a.iter().cloned().collect();        
    }
    Arc::get_mut(a).unwrap()
}
```

Reply:
And let's remove `_arc` from the name, while we are at it: it's unliccely we'll evere extend this to RC